pandas/core/reshape/merge.py

```diff
@@ -1596,6 +1623,25 @@ def _validate(self, validate: str) -> None:
             )
 
 
+def _semi_helper(
```

Comment:
Why do we need this function? Isn't isin already implemented for all of the types that would be used in a join already?

Reply:
At first glance, it looks to me as if .isin() is not used for the other join methods because we append additional columns from the other data frame there. But I could be wrong with my assumption, as I am not familiar with the "low level" functionality of pandas.

In the other pull request it was stated that we could stick with that solution:

"I don't think this should go in that direction unless there already is a viable alternative; to my knowledge there isn't. As I remarked above, while this performance is less than left-join on my columns, I feel it's sufficient for the feature."